dan_layer/core/src/workers/pacemaker_worker.rs

```diff
@@ -0,0 +1,270 @@
+    pub fn spawn(
+        rx_start_signal: Receiver<T>,
+        rx_shutdown_signal: Receiver<T>,
+        tx_waiter_status: Sender<(T, PacemakerWaitStatus)>,
+        max_timeout: u64,
+        shutdown: ShutdownSignal,
+    ) -> JoinHandle<Result<(), HotStuffError>> {
+        let pacemaker = Self::new(rx_start_signal, rx_shutdown_signal, tx_waiter_status, max_timeout);
+        tokio::spawn(pacemaker.run(shutdown))
+    }
+
+    fn new(
+        rx_start_signal: Receiver<T>,
+        rx_shutdown_signal: Receiver<T>,
+        tx_waiter_status: Sender<(T, PacemakerWaitStatus)>,
+        max_timeout: u64,
+    ) -> Self {
+        Self {
+            rx_start_signal,
+            rx_shutdown_signal,
+            tx_inner_map: HashMap::new(),
+            tx_waiter_status,
+            max_timeout,
+        }
+    }
+
+    pub async fn run(mut self, mut shutdown: ShutdownSignal) -> Result<(), HotStuffError> {
+        let max_timeout = self.max_timeout;
+        loop {
+            tokio::select! {
+                msg = self.rx_start_signal.recv() => {
+                    if let Some(wait_over) = msg {
+                        let (tx, mut rx) = channel::<bool>(1);
+                        self.tx_inner_map.insert(wait_over, tx);
+                        info!(
+                            target: LOG_TARGET,
+                            "Received start wait signal for value: {:?}", wait_over
+                        );
+                        let tx_waiter_status = self.tx_waiter_status.clone();
+                        let _join = tokio::spawn(async move {
+                            tokio::select! {
+                                _ = tokio::time::sleep(Duration::from_millis(max_timeout)) => {
+                                    if let Err(e) = send_timeout_message(wait_over, tx_waiter_status).await {
+                                        error!(target: LOG_TARGET, "failed to send timeout status message for value = {:?} with error = {}", wait_over, e);
+                                    }
+                                },
+                                _ = rx.recv() => {
+                                    info!("The wait signal for wait_over = {:?} has been shutted down", wait_over);
+                                }
+                            }
+                        });
+                    }
+                },
+                msg = self.rx_shutdown_signal.recv() => {
+                    if let Some(wait_over) = msg {
+                        if let Some(tx) = self.tx_inner_map.get(&wait_over) {
+                            tx.send(true).await.unwrap();
+                            // remove the entry from the tx_inner_map
+                            self.tx_inner_map.remove(&wait_over);
+                        }
+                    }
+                },
+                _ = shutdown.wait() => {
+                    info!("Shutting down pacemaker service..");
```

Comment:
I'm not sure if it's necessary, but maybe you should trigger all shutdowns. So nothing is left behind running.